File: VB/PatchCriteriaExample/Filtering/CriteriaPatcherSkipProperties.vb
Imports DevExpress.Data.Filtering
Imports System.Collections.Generic
Imports DevExpress.Data.Filtering.Helpers

Namespace PatchCriteriaExample.Filtering
    Public Class CriteriaPatcherSkipProperties
        Inherits ClientCriteriaLazyPatcherBase.AggragatesCommonProcessingBase

        Private ReadOnly PropertiesToremove As IList(Of String)

        Private Sub New(ByVal propertiesToRemove As String)
            Me.PropertiesToremove = New List(Of String)(propertiesToRemove.Split(","c))
        End Sub

        Public Shared Function Patch(ByVal source As CriteriaOperator, ByVal propertiesToRemove As String) As CriteriaOperator
            Return (New CriteriaPatcherSkipProperties(propertiesToRemove)).Process(source)
        End Function

        Private Shared Function IsNothing(theOperator As CriteriaOperator) As Boolean
            Return Object.ReferenceEquals(theOperator, Nothing)
        End Function

        Private Shared Function RemoveEmptyOperands(ByVal source As CriteriaOperatorCollection) As CriteriaOperatorCollection
            Dim result As New CriteriaOperatorCollection()
            For Each operand As CriteriaOperator In source
                If Not IsNothing(operand) Then
                    result.Add(operand)
                End If
            Next operand
            Return result
        End Function

        Public Overrides Function Visit(ByVal theOperand As OperandProperty) As CriteriaOperator
            If PropertiesToremove.Contains(theOperand.PropertyName) Then
                Return Nothing
            End If
            Return theOperand
        End Function

        Public Overrides Function Visit(ByVal theOperand As AggregateOperand) As CriteriaOperator
            Dim collectionProperty As CriteriaOperator = Visit(theOperand.CollectionProperty)
            If IsNothing(collectionProperty) Then
                Return Nothing
            End If
            Dim patched As CriteriaOperator = MyBase.Visit(theOperand)
            If Object.ReferenceEquals(theOperand, patched) Then
                Return theOperand
            End If
            Return Nothing
        End Function
        Public Overrides Function Visit(ByVal theOperator As BetweenOperator) As CriteriaOperator
            theOperator = CType(MyBase.Visit(theOperator), BetweenOperator)
            If IsNothing(theOperator.BeginExpression) OrElse IsNothing(theOperator.EndExpression) OrElse IsNothing(theOperator.TestExpression) Then
                Return Nothing
            End If
            Return theOperator
        End Function

        Public Overrides Function Visit(ByVal theOperator As BinaryOperator) As CriteriaOperator
            theOperator = CType(MyBase.Visit(theOperator), BinaryOperator)
            If IsNothing(theOperator.LeftOperand) OrElse IsNothing(theOperator.RightOperand) Then
                Return Nothing
            End If
            Return theOperator
        End Function

        Public Overrides Function Visit(ByVal theOperator As FunctionOperator) As CriteriaOperator
            Dim result = CType(MyBase.Visit(theOperator), FunctionOperator)
            If Not Object.ReferenceEquals(theOperator, result) Then
                Return Nothing
            End If
            Return result
        End Function

        Public Overrides Function Visit(ByVal theOperator As InOperator) As CriteriaOperator
            Dim result = CType(MyBase.Visit(theOperator), InOperator)
            If IsNothing(result.LeftOperand) Then
                Return Nothing
            End If
            If Object.ReferenceEquals(theOperator.Operands, result.Operands) Then
                Return theOperator
            End If
            Dim filteredOperands As CriteriaOperatorCollection = RemoveEmptyOperands(result.Operands)
            If filteredOperands.Count = 0 Then
                Return Nothing
            End If
            If filteredOperands.Count = 1 Then
                Return New BinaryOperator(theOperator.LeftOperand, filteredOperands(0), BinaryOperatorType.Equal)
            End If
            Return New InOperator(theOperator.LeftOperand, filteredOperands)
        End Function

        Public Overrides Function Visit(ByVal theOperator As UnaryOperator) As CriteriaOperator
            theOperator = CType(MyBase.Visit(theOperator), UnaryOperator)
            If IsNothing(theOperator.Operand) Then
                Return Nothing
            End If
            Return theOperator
        End Function
    End Class
End Namespace

File: CS/PatchCriteriaExample/Filtering/CriteriaPatcherSkipProperties.cs
using DevExpress.Data.Filtering;
using DevExpress.Data.Filtering.Helpers;
using System.Collections.Generic;

namespace PatchCriteriaExample.Filtering {
    public class CriteriaPatcherSkipProperties : ClientCriteriaLazyPatcherBase.AggregatesCommonProcessingBase {
        private readonly IList<string> PropertiesToremove;

        private CriteriaPatcherSkipProperties(string propertiesToRemove) {
            this.PropertiesToremove = new List<string>(propertiesToRemove.Split(','));
        }

        public static CriteriaOperator Patch(CriteriaOperator source, string propertiesToRemove) {
            return new CriteriaPatcherSkipProperties(propertiesToRemove).Process(source);
        }

        private static bool IsNull(CriteriaOperator theOperator) {
            return object.ReferenceEquals(theOperator, null);
        }

        public override CriteriaOperator Visit(OperandProperty theOperand) {
            if(PropertiesToremove.Contains(theOperand.PropertyName)) return null;
            return theOperand;
        }

        public override CriteriaOperator Visit(AggregateOperand theOperand) {
            CriteriaOperator collectionProperty = Visit(theOperand.CollectionProperty);
            if(IsNull(collectionProperty)) return null;
            CriteriaOperator patched =  base.Visit(theOperand);
            if(object.ReferenceEquals(theOperand, patched)) return theOperand;
            return null;
        }

        public override CriteriaOperator Visit(BetweenOperator theOperator) {
             theOperator = (BetweenOperator)base.Visit(theOperator);
             if(IsNull(theOperator.BeginExpression) || IsNull(theOperator.EndExpression)
                 || IsNull(theOperator.TestExpression))
                 return null;
             return theOperator;
        }

        public override CriteriaOperator Visit(BinaryOperator theOperator) {
            theOperator = (BinaryOperator)base.Visit(theOperator);
            if(IsNull(theOperator.LeftOperand) || IsNull(theOperator.RightOperand))
                return null;
            return theOperator;
        }

        public override CriteriaOperator Visit(FunctionOperator theOperator) {
            var result = (FunctionOperator)base.Visit(theOperator);
            if(!object.ReferenceEquals(theOperator, result)) return null;
            return result;
        }

        public override CriteriaOperator Visit(InOperator theOperator) {
            var result = (InOperator)base.Visit(theOperator);
            if(IsNull(result.LeftOperand)) return null;
            if(object.ReferenceEquals(theOperator.Operands, result.Operands)) return theOperator;
            CriteriaOperatorCollection filteredOperands = RemoveEmptyOperands(result.Operands);
            if(filteredOperands.Count == 0) return null;
            if(filteredOperands.Count == 1)
                return new BinaryOperator(theOperator.LeftOperand, filteredOperands[0], BinaryOperatorType.Equal);
            return new InOperator(theOperator.LeftOperand, filteredOperands);
        }

        public override CriteriaOperator Visit(UnaryOperator theOperator) {
            theOperator = (UnaryOperator)base.Visit(theOperator);
            if(IsNull(theOperator.Operand)) return null;
            return theOperator;
        }

        private static CriteriaOperatorCollection RemoveEmptyOperands(CriteriaOperatorCollection source) {
            CriteriaOperatorCollection result = new CriteriaOperatorCollection();
            foreach(CriteriaOperator operand in source)
                if(!IsNull(operand)) result.Add(operand);
            return result;
        }
    }
}
